### ISSUE TITLE
Unable to set screen options of child routes when a _layout contains a Slot

### ISSUE THREAD

**Issue by mattwilliams0592** [expo/expo#27894]
### Minimal reproducible example

https://github.com/mattwilliams0592/expo-router-slot-issue

### Which package manager are you using? (Yarn is recommended)

npm

### If the issue is web-related, please select the bundler (`web.bundler` in the `app.json`)

None

### Summary

When using a Slot within a _layout route, you cannot set screen options of child routes unless you make use of the parent navigation, which I don't think is the intended usage. This may be the wrong terminology, but it seems that routes within Slots are losing context of/relation to their nearest parent.

**Notable files:**
- /app/without-layout: This route **does not** have a _layout, and setting screen options works as intended.
- /app/with-layout: This route **has** a _layout with a Slot, and setting screen options does not work.
- /app/with-layout-get-parent: This route **has** a _layout with a Slot, and setting screen options works, but only by using the parent navigation.

### Environment

  expo-env-info 1.2.0 environment info:
    System:
      OS: macOS 14.4
      Shell: 5.9 - /bin/zsh
    Binaries:
      Node: 18.19.0 - ~/.nvm/versions/node/v18.19.0/bin/node
      npm: 10.2.4 - ~/.nvm/versions/node/v18.19.0/bin/npm
    SDKs:
      iOS SDK:
        Platforms: DriverKit 23.0, iOS 17.0, macOS 14.0, tvOS 17.0, watchOS 10.0
    IDEs:
      Xcode: 15.0.1/15A507 - /usr/bin/xcodebuild
    npmPackages:
      expo: ~50.0.14 => 50.0.14
      expo-router: ~3.4.8 => 3.4.8
      react: 18.2.0 => 18.2.0
      react-dom: 18.2.0 => 18.2.0
      react-native: 0.73.6 => 0.73.6
      react-native-web: ~0.19.6 => 0.19.10
    npmGlobalPackages:
      eas-cli: 7.0.0
    Expo Workflow: managed

**Comment by marklawlor** [comment]
`<Slot />` don't lose the context of their parent, they are a new node in the navigation hierarchy so they form their own context. `<Stack.screen />` simply passes options to the current navigator and as the `<Slot />` doesn't have a visual appearance, nothing happens. This is also why `useNavigation().getParent()` works (`useNavigation` return the `<Slot />`, `getParent` return the `<Stack />`)

While we are working to make `<Slot />` easier to use, setting screen options like this is problematic for Expo Router and something we might remove in the future. For example, it doesn't work well with static rendering and will use optimisation opt-outs of future features (e.g React Server Components). 

Another way to implement a dynamic title is provide a custom component to the `Stack.Screen` within your layout file

```
<Stack.Navigator>
  <Stack.Screen
    name="with-layout"
    options={{ headerTitle: (props) => <Title {...props} /> }}
  />
</Stack.Navigator>
```

**Comment by fmendez89** [comment]
> Thanks for the prompt response @marklawlor, much appreciated. I was under the impression that a `<Slot />` functioned more along the lines of standard `children`, but I understand now with regard to the new node in the navigation hierarchy you mentioned.
> 
> In my application, screen options are set further down the tree within a custom `<Screen />` component that accepts a `title` prop. This component handles the presentation of different navigational UI's based on screen size and platform. I use a singular `<Stack />` navigation throughout the app, so my use case allows me to recursively get the root (parent) navigation when necessary.
> 
> By the way, loving Expo Router and the possibilities it provides; keep up the great work.

Hi @mattwilliams0592, did you find a solution? My use case is the same, I tried different approaches, but without any luck.

**Comment by mattwilliams0592** [comment]
Hi @fmendez89, yes and no. The solution I came to was to avoid the use of `<Slot />` entirely. I intended to use them to add safeguards and context checks in `_layout` files, but because of the way they work with regard to the new node in the navigation hierarchy, it made it quite difficult.

What I ended up doing is only using `_layout` files when absolutely necessary and doing more context checks in my root `_layout` file. For example, to ensure an unauthenticated user hasn't navigated to an authenticated route, I do a check against the segment of the app they're in (`const baseSegment = useSegments()[0]`) and their authentication state.

When I do need to make use of a `_layout` file then I'll render another `<Stack />` or `<Tabs />`. This `_layout` file ends up looking something like the example below. The key here is the hiding of the parent navigation header, but just note that you'll need to manually add a back button to the subroute as well:

```
export default () => {
  const navigation = useNavigation();
  const someContext = useSomeContext();

  React.useEffect(() => {
    // Hide the header of the parent stack navigator
    navigation.setOptions({ headerShown: false });
  }, []);

  if (!someContext.isValid) {
    return <Redirect href="/somewhere-else" />;
  }

  return <Stack />;
}
```

It's not a fantastic DX, but hopefully one day there will be a better solution.

**Comment by fmendez89** [comment]
Thanks @mattwilliams0592 
I actually tried using the setOptions, but within a useLayoutEffect. The result was almost identical, but the DX was awful, so I removed the Slot and used a different approach because it became a mess to maintain. The only Slot in the app is at the root, that way the navigation is perfect in my case.

Thanks for the explanation anyway, I really thought Slot was behaving like children as the documentation states.
